From ad439978ddaedabded3d0f97d69df6898656a1a5 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Tue, 6 Oct 2026 06:33:54 +0200
Subject: [PATCH] TFEL: tensor accessors usable in device code

The tensor view of a symmetric tensor, used by the operations mixing
tensors and symmetric tensors (for instance the product of the
deformation gradient by the second Piola-Kirchhoff stress), and the
matrix-like accessors of tensors were host functions, because they throw
an exception for an invalid index. clang rejects them in device code;
nvcc accepted the constexpr ones with --expt-relaxed-constexpr.

The new internals::reportInvalidTensorIndex function throws the
exception on the host and returns a NaN value in device code, as
TinyMatrixInvert does for singular matrices.
---
 include/TFEL/Math/Tensor/TensorConcept.hxx    | 25 ++++++++++++++++++-
 include/TFEL/Math/Tensor/TensorConcept.ixx    |  8 +++---
 .../Math/Tensor/TensorViewFromStensor.hxx     | 18 ++++++++-----
 include/TFEL/Math/Tensor/tensor.ixx           |  5 ++--
 include/TFEL/Math/tensor.hxx                  |  2 +-
 5 files changed, 45 insertions(+), 13 deletions(-)

diff --git a/include/TFEL/Math/Tensor/TensorConcept.hxx b/include/TFEL/Math/Tensor/TensorConcept.hxx
index d0fb2a8b8..00b3cbc3d 100644
--- a/include/TFEL/Math/Tensor/TensorConcept.hxx
+++ b/include/TFEL/Math/Tensor/TensorConcept.hxx
@@ -14,6 +14,7 @@
 #ifndef LIB_TFEL_MATH_TENSORCONCEPT_HXX
 #define LIB_TFEL_MATH_TENSORCONCEPT_HXX 1
 
+#include <limits>
 #include <stdexcept>
 #include <type_traits>
 #include "TFEL/Config/TFELConfig.hxx"
@@ -94,6 +95,28 @@ namespace tfel::math {
     ~TensorInvalidIndexesException() noexcept override;
   };  // end of struct TensorInvalidIndexesException
 
+  namespace internals {
+
+    /*!
+     * \brief report an invalid index in an access to the values of a tensor
+     *
+     * On the host, an exception of the given type is thrown. Device code
+     * can't throw exceptions: a NaN value is returned.
+     *
+     * \tparam Exception: type of the exception
+     * \tparam ValueType: type of the values of the tensor
+     */
+    template <typename Exception, typename ValueType>
+    TFEL_HOST_DEVICE constexpr ValueType reportInvalidTensorIndex() {
+#ifdef TFEL_DEVICE_CODE
+      return ValueType(std::numeric_limits<base_type<ValueType>>::quiet_NaN());
+#else  /* TFEL_DEVICE_CODE */
+      throw(Exception());
+#endif /* TFEL_DEVICE_CODE */
+    }  // end of reportInvalidTensorIndex
+
+  }  // end of namespace internals
+
   /*!
    * \brief exception thrown if an std::initializer_list with no
    * appropriate size is affected to a tensor.
@@ -120,7 +143,7 @@ namespace tfel::math {
      * \param[in] i: row number
      * \param[in] j: column number
      */
-    [[nodiscard]] TFEL_HOST numeric_type<T> operator()(
+    [[nodiscard]] TFEL_HOST_DEVICE numeric_type<T> operator()(
         const unsigned short, const unsigned short) const;
   };
 
diff --git a/include/TFEL/Math/Tensor/TensorConcept.ixx b/include/TFEL/Math/Tensor/TensorConcept.ixx
index 6e054151a..47727f881 100644
--- a/include/TFEL/Math/Tensor/TensorConcept.ixx
+++ b/include/TFEL/Math/Tensor/TensorConcept.ixx
@@ -74,7 +74,8 @@ namespace tfel::math::internals {
       } else if ((i == 2) && (j == 1)) {
         return t(8);
       }
-      throw(TensorInvalidIndexesException());
+      return reportInvalidTensorIndex<TensorInvalidIndexesException,
+                                      numeric_type<TensorType>>();
     }
   };
 
@@ -83,11 +84,12 @@ namespace tfel::math::internals {
 namespace tfel::math {
 
   template <typename T>
-  TFEL_HOST numeric_type<T> TensorConceptBase<T>::operator()(
+  TFEL_HOST_DEVICE numeric_type<T> TensorConceptBase<T>::operator()(
       const unsigned short i, const unsigned short j) const {
     using tfel::math::internals::TensorConceptMatrixAccessOperator;
     if ((i > 2) || (j > 2)) {
-      throw(TensorInvalidIndexesException());
+      return internals::reportInvalidTensorIndex<TensorInvalidIndexesException,
+                                                 numeric_type<T>>();
     }
     return TensorConceptMatrixAccessOperator<getSpaceDimension<T>()>::exe(
         static_cast<const T&>(*this), i, j);
diff --git a/include/TFEL/Math/Tensor/TensorViewFromStensor.hxx b/include/TFEL/Math/Tensor/TensorViewFromStensor.hxx
index 517b92a2b..740ad6b46 100644
--- a/include/TFEL/Math/Tensor/TensorViewFromStensor.hxx
+++ b/include/TFEL/Math/Tensor/TensorViewFromStensor.hxx
@@ -45,11 +45,13 @@ namespace tfel::math {
     TFEL_HOST_DEVICE constexpr Expr& operator=(const Expr&) = default;
     TFEL_HOST_DEVICE constexpr Expr& operator=(Expr&&) = default;
 
-    [[nodiscard]] TFEL_HOST ValueType operator[](const unsigned short i) const {
+    [[nodiscard]] TFEL_HOST_DEVICE constexpr ValueType operator[](
+        const unsigned short i) const {
       return this->operator()(i);
     }
 
-    [[nodiscard]] TFEL_HOST ValueType operator()(const unsigned short i) const {
+    [[nodiscard]] TFEL_HOST_DEVICE constexpr ValueType operator()(
+        const unsigned short i) const {
       constexpr auto icste = Cste<ValueType>::isqrt2;
       switch (i) {
         case 0:
@@ -61,25 +63,29 @@ namespace tfel::math {
         case 3:
         case 4:
           if (traits::dime < 2) {
-            throw(TensorInvalidIndexException());
+            return internals::reportInvalidTensorIndex<
+                TensorInvalidIndexException, ValueType>();
           }
           return this->s(3) * icste;
         case 5:
         case 6:
           if (traits::dime < 3) {
-            throw(TensorInvalidIndexException());
+            return internals::reportInvalidTensorIndex<
+                TensorInvalidIndexException, ValueType>();
           }
           return this->s(4) * icste;
         case 7:
         case 8:
           if (traits::dime < 3) {
-            throw(TensorInvalidIndexException());
+            return internals::reportInvalidTensorIndex<
+                TensorInvalidIndexException, ValueType>();
           }
           return this->s(5) * icste;
         default:
           break;
       }
-      throw(TensorInvalidIndexException());
+      return internals::reportInvalidTensorIndex<TensorInvalidIndexException,
+                                                 ValueType>();
     }  // end of operator() const
 
    protected:
diff --git a/include/TFEL/Math/Tensor/tensor.ixx b/include/TFEL/Math/Tensor/tensor.ixx
index 55e395a40..bf29879b2 100644
--- a/include/TFEL/Math/Tensor/tensor.ixx
+++ b/include/TFEL/Math/Tensor/tensor.ixx
@@ -61,12 +61,13 @@ namespace tfel::math {
   }
 
   template <unsigned short N, typename ValueType>
-  TFEL_HOST constexpr ValueType tensor<N, ValueType>::operator()(
+  TFEL_HOST_DEVICE constexpr ValueType tensor<N, ValueType>::operator()(
       const typename tensor<N, ValueType>::size_type i,
       const typename tensor<N, ValueType>::size_type j) const {
     static_assert((N == 1) || (N == 2) || (N == 3), "invalid space dimension");
     if ((i > 2) || (j > 2)) {
-      throw(TensorInvalidIndexesException());
+      return internals::reportInvalidTensorIndex<TensorInvalidIndexesException,
+                                                 ValueType>();
     }
     if constexpr (N == 1) {
       return (i == j) ? (*this)(i) : ValueType(0);
diff --git a/include/TFEL/Math/tensor.hxx b/include/TFEL/Math/tensor.hxx
index 191d2cf3d..122f46873 100644
--- a/include/TFEL/Math/tensor.hxx
+++ b/include/TFEL/Math/tensor.hxx
@@ -128,7 +128,7 @@ namespace tfel::math {
      * \param[in] i: row number
      * \param[in] j: column number
      */
-    TFEL_HOST constexpr ValueType operator()(
+    TFEL_HOST_DEVICE constexpr ValueType operator()(
         const typename tensor<N, ValueType>::size_type,
         const typename tensor<N, ValueType>::size_type) const;
     //! \brief write to an external memory location
-- 
2.43.0

